ui/test/nuxeo-document-layout.test.js

```diff
@@ -0,0 +1,309 @@
+    );
+    expect(documentLayout.$.layout.$.error.url).to.equal(documentLayout._href);
+  };
+
+  test('Should display an error when the layout is not found', async () => {
+    documentLayout = await buildLayout(buildDoc(), 'metadata');
+    // we have no such view layout, so it will result in 404
+    assertNotFound();
+  });
+
+  test('Should load a layout when a document and a layout are provided', async () => {
+    documentLayout = await buildLayout();
+    // check the layout was loaded
+    expect(documentLayout.element).to.exist;
+    expect(documentLayout.element.tagName).to.equal('NUXEO-TEST-EDIT-LAYOUT');
+    // assert layout integrity
+    expect(documentLayout.element.document).to.equal(documentLayout.document);
+    const { children } = documentLayout.element.shadowRoot;
+    expect(children).to.have.lengthOf(3);
+    expect(children[0].tagName).to.equal('STYLE');
+    expect(children[1].tagName).to.equal('NUXEO-INPUT');
+    expect(children[2].tagName).to.equal('NUXEO-TEXTAREA');
+  });
+
+  test('Should fire a "document-layout-changed" event when the layout is loaded', async () => {
+    // we need to ad a listener for this event before hand, because we do not yet have the layout stamped,
+    // and after `buildLayout` it can be too late to catch the event.
+    const layoutChanged = new Promise((resolve) => {
+      document.addEventListener('document-layout-changed', (e) => resolve(e));
+    });
+    documentLayout = await buildLayout();
+    // check that the event was fired from the document layout
+    const event = await layoutChanged;
+    expect(event.target).to.equal(documentLayout);
+    // ensure that the event bubbles and is composed
+    expect(event.bubbles).to.be.true;
+    expect(event.composed).to.be.true;
+    // verify event details
+    expect(event.detail).to.exist;
+    expect(event.detail.element).to.equal(documentLayout.element);
+    expect(event.detail.layout).to.equal(documentLayout.layout);
+  });
+
+  test('Should reload when layout name changes', async () => {
+    documentLayout = await buildLayout();
+    expect(documentLayout.element).to.exist;
+    expect(documentLayout.element.tagName).to.equal('NUXEO-TEST-EDIT-LAYOUT');
+    // check that the layout updates when the layout name changes
+    documentLayout.layout = 'view';
+    const event = await waitForEvent(documentLayout, 'document-layout-changed');
+    expect(event.detail).to.exist;
+    const { element } = event.detail;
+    expect(element).to.exist;
+    expect(element.tagName).to.equal('NUXEO-TEST-VIEW-LAYOUT');
+    // assert layout integrity
+    const { children } = element.shadowRoot;
+    expect(children).to.have.lengthOf(3);
+    expect(children[0].tagName).to.equal('STYLE');
+    expect(children[1].tagName).to.equal('H1');
+    expect(children[2].tagName).to.equal('DIV');
+  });
+
+  test('Should reload when document changes', async () => {
```

Comment:
And what happens when we change document but for the same doc type? Is it an interesting test?

Reply:
Not sure. This reload is re-triggered when the document changes, which generates a new href. I have no reason to believe that we'd need to go that far as long as we're generating hrefs correctly. So, perhaps this means that we should be testing more `hrefFunction` as well?

Reply:
Added a test for the `hrefFunction`.